src/Messages.jl

```diff
@@ -0,0 +1,555 @@
+[`HTTP.Messages.readstartline!`](@ref).
+
+The `read` methods rely on [`HTTP.IOExtras.unread!`](@ref) to push excess
+data back to the input stream.
+
+
+### Headers
+
+Headers are represented by `Vector{Pair{String,String}}`. As compared to
+`Dict{String,String}` this allows [repeated header fields and preservation of
+order](https://tools.ietf.org/html/rfc7230#section-3.2.2).
+
+Header values can be accessed by name using
+[`HTTP.Messages.header`](@ref) and
+[`HTTP.Messages.setheader`](@ref) (case-insensitive).
+
+The [`HTTP.Messages.appendheader`](@ref) function handles combining
+multi-line values, repeated header fields and special handling of
+multiple `Set-Cookie` headers.
+
+### Bodies
+
+The `HTTP.Message` structs represent the Message Body as `Vector{UInt8}`.
+
+Streaming of request and response bodies is handled by the
+[`HTTP.StreamLayer`](@ref) and the [`HTTP.Stream`](@ref) `<: IO` stream.
+"""
+
+
+module Messages
+
+export Message, Request, Response,
+       reset!,
+       iserror, isredirect, ischunked, issafe, isidempotent,
+       header, hasheader, setheader, defaultheader, appendheader,
+       mkheaders, readheaders, headerscomplete, readtrailers, writeheaders,
+       readstartline!, writestartline,
+       bodylength, unknown_length
+
+if VERSION > v"0.7.0-DEV.2338"
+using Unicode
+end
+
+import ..HTTP
+
+using ..Pairs
+using ..IOExtras
+using ..Parsers
+import ..Parsers
+import ..Parsers: headerscomplete, reset!
+
+const unknown_length = typemax(Int64)
+
+
+abstract type Message end
+
+"""
+    Response <: Message
+
+Represents a HTTP Response Message.
+
+- `version::VersionNumber`
+- `status::Int16`
+- `headers::Vector{Pair{String,String}}`
+- `body::Vector{UInt8}`
+- `request`, the `Request` that yielded this `Response`.
+"""
+
+mutable struct Response <: Message
+    version::VersionNumber
+    status::Int16
+    headers::Headers
+    body::Vector{UInt8}
+    request::Message
+
+    function Response(status::Int=0, headers=[]; body=UInt8[], request=nothing)
+        r = new()
+        r.version = v"1.1"
+        r.status = status
+        r.headers = mkheaders(headers)
+        r.body = body
+        if request != nothing
+            r.request = request
+        end
+        return r
+    end
+end
+
+Response(bytes) = parse(Response, bytes)
+
+function reset!(r::Response)
+    r.version = v"1.1"
+    r.status = 0
+    if !isempty(r.headers)
+        empty!(r.headers)
+    end
+    if !isempty(r.body)
+        empty!(r.body)
+    end
+end
+
+
+"""
+    Request <: Message
+
+Represents a HTTP Request Message.
+
+- `method::String`
+- `uri::String`
+- `version::VersionNumber`
+- `headers::Vector{Pair{String,String}}`
+- `body::Vector{UInt8}`
+- `response`, the `Response` to this `Request`
+- `parent`, the `Response` (if any) that led to this request
+  (e.g. in the case of a redirect).
+"""
+
+mutable struct Request <: Message
+    method::String
+    uri::String
```

Comment:
why not store this as a `URI`? it seems like it will be annoying on the server-side to always have to parse the uri before being able to get any parts out of it.

Reply:
And `method::Method`?

I made a similar change a while back. It should probably be `method::Method` and `uri::Uri` in the struct definition but we can allow for strings in a separate constructor. Maybe?

Reply:
Having reviewed the full PR now, I'm actually ok w/ this change. I think @samoconnor probably realized that it simplifies some internals by just leaving them strings, and I'm ok w/ that. There's no super compelling reason to make them special structs.

Reply:
I think the principals of decoupling and separation of concerns apply here. If the Messages modules can do its job while treating `method` and `uri` as just being opaque chunks of string, then they should not have a more specific type. The less knowledge mode module has of another the better. The smaller the interface surface area the better.

Also, laziness is good for performance. Just say we parse every url string into a `URI` object. Now suppose there is some use case where a whole bunch of Request objects get constructed, used for some purpose, then discarded,  but the code in question never needs to access the `URI.query()`. All the time spent parsing strings into `URI` objects was wasted.